Fix Static File Creation Issue

smg6511 opened the pull request:
### What does it do?

Corrected the file path reference used to create new static element files.

### Why is it needed?

Two connected problems are resolved here:
1. The logging issue referenced in issue #16005
2. Without this fix, a duplicate file (which is empty) is produced within the selected media source (but having the full server path). For example, if choosing the "Filesystem" media source and creating a static chunk at "assets/chunks/static-chunk-1.html", two files will be created (assuming an example server root path of "/var/www/site"):
-  assets/chunks/static-chunk-1.html (as expected)
-  assets/chunks/var/www/site/assets/chunks/static-chunk-1.html (empty duplicate at wrong location)

### How to Test

1. To verify the second problem above, create a static element or two on any other install/branch of 3.x. Take a look at your logs and the directories/files that get created.
2. Load _this_ PR branch and create one of each element with a static file source specified. For each, be sure to enter a little text/code into the element's code field to verify its content is actually saved to the target file. 
3. Monitor your manager error log to verify that entries do not get created when a static element is successfully created and saved. 
4. Verify in your file structure that there is no duplication of the static file in another area of the file system (see item 2 above).
5. Although probably not necessary (because I don't believe this code intersects with the parsing end of things), create a page that makes use of each static element to ensure the output is as expected.

### Additional Note

This PR does not seek to fix either of these two issues:
1. If you specify a file type that is not listed in the `upload_files` system setting  you will be unable to save the static element. A resolution for this, perhaps by implementing a new system setting for allowable static types and/or the ability to specify allowable types on a media source basis, will be proposed in a separate PR.
2. Selecting "None" as the media source currently results in unexpected behavior and absolute paths do not work (see also #15927). This will be addressed in a separate PR as well.

### Related issue(s)/PR(s)
Resolves #16005

Commit:
Fix Static File Path Issue

Ruslan-Aleev commented on core/lexicon/en/file.inc.php:
```diff
@@ -73,6 +73,7 @@
 $_lang['file_folder_remove'] = 'Delete Directory';
 $_lang['file_folder_remove_confirm'] = 'Are you sure you want to delete the directory: "[[+directory]]"?<br />This could potentially break your website.';
 $_lang['file_folder_rename'] = 'Rename Directory';
+$_lang['file_info_metadata_new'] = "Unable to access metaData for the file “[[+file]]” because it is for a new element and has not been created yet. An attempt will be made to create the file automatically.";
```
You are using the wrong quotes (both for the whole line and for the placeholder [[+file]]).
I don't think it's important, but it's better to keep them in the same style (since crowdin sometimes behaves strangely) :)

smg6511 replied:
Yeah, I'm picky about using single quotes everywhere possible ... missed that here, will change! On the other quotes, they are proper quotation marks for text that I've used elsewhere (using the utf-8 curly quotation marks instead of html entities).

Ruslan-Aleev replied:
It is better to lead to one style.

smg6511 replied:
Made changes to the lex file. Should be all consistent now.

Commit:
Update file.inc.php

Making quotation style consistent and changing xhtml-style br tags to html5

Jako commented on core/src/Revolution/Sources/modMediaSource.php:
```diff
@@ -545,21 +545,25 @@ public function getObjectsInContainer($path)
+            if (!$isNew) {
+                $this->xpdo->log(modX::LOG_LEVEL_ERROR, $e->getMessage());
+            } else {
+                $this->xpdo->log(modX::LOG_LEVEL_ERROR, "\r\n\t" . $this->xpdo->lexicon('file_info_metadata_new', ['file' => $path]));
```
Why do yo create a new line and a tab indent here?

smg6511 replied:
It's just easier to read the part of the output that's most important IMO.

Jako commented on core/lexicon/en/file.inc.php:
```diff
@@ -71,8 +71,9 @@
 $_lang['file_folder_parent_desc'] = 'Relative to the basePath of the media source';
 $_lang['file_folder_remove'] = 'Delete Directory';
-$_lang['file_folder_remove_confirm'] = 'Are you sure you want to delete the directory: "[[+directory]]"?<br />This could potentially break your website.';
+$_lang['file_folder_remove_confirm'] = 'Are you sure you want to delete the directory: “[[+directory]]”?<br>This could potentially break your website.';
```
I am fine using the right typography here, but the whole lexicons are different with double quotes.

smg6511 replied:
Yes, I know. But you've got to start somewhere. I can address the other places in a different PR. I don't see it as a problem to update them over time, as it's not a highly-visible change.

Jako commented on core/lexicon/en/file.inc.php:
```diff
@@ -117,7 +118,7 @@
 $_lang['upf_close_desc'] = 'Close the window.';
 $_lang['upf_err_failed'] = 'Server is unavailable or internal server error occured.';
-$_lang['upf_err_filetype'] = 'Selected file extension isn\'t permitted.<br/>Please select files with following extensions: {1}';
+$_lang['upf_err_filetype'] = 'Selected file extension isn’t permitted.<br>Please select files with following extensions: {1}';
```
I am fine using the right typography here, but the whole lexicons are different with single quotes.

smg6511 replied:
Resolving here, as we can discuss this above re double quotes.

Ruslan-Aleev replied:
It is better to return it as it was, because all changed strings will overwrite translations in Crowdin.
After a few of these iterations, MODX will lose all translations.
p.s. And for the new line, leave it in the old style: firstly, it will look in one style, and secondly, there will be no confusion with quotes.

Commit:
Reversion of select formatting changes

Walking back changes, per PR reviewers’ requests

smg6511 commented:
@Jako @Ruslan-Aleev - Ok guys, I walked back the small formatting changes and will let that go _for now_. Please take a look at this PR for its functionality, soon if you can ;-) 

This is the first of three (independent) PRs I'm working on the static elements issues that I think are pretty critical to the upcoming public release (as @JoshuaLuckers had indicated recently on Slack). The next will tackle the specification of static files that don't use a media source (and rely on an absolute path); this simply does not work currently and is a little more involved because of the move to Flysystem (for file handling). The last addresses the duplicate functionality (which has a bunch of problems as well, though maybe less critical).

JoshuaLuckers commented on core/src/Revolution/Sources/modMediaSource.php:
```diff
@@ -545,21 +545,25 @@ public function getObjectsInContainer($path)
+            if (!$isNew) {
+                $this->xpdo->log(modX::LOG_LEVEL_ERROR, $e->getMessage());
+            } else {
+                $this->xpdo->log(modX::LOG_LEVEL_ERROR, $this->xpdo->lexicon('file_info_metadata_new', ['file' => $path]));
```
Isn't the message we log debug information instead of an error?

> Unable to access metaData for the file “test.php” because it is for a new element and has not been created yet. An attempt will be made to create the file automatically.

smg6511 replied:
It is, yes. A couple things here: I thought it might be helpful for it to be noticable for testing. I did have in mind knocking it down to INFO in the end. Also, I see this disappearing at some point, hopefully soon. I looked very carefully at how the file handling process works now and find it rather convoluted. Why is getMetaData _any_ part of checking whether a file exists or not? This method should strictly be used for what it implies its use is: Fetching the metadata of an existing file.

Mark-H commented on core/src/Revolution/Sources/modMediaSource.php:
```diff
@@ -545,21 +545,25 @@ public function getObjectsInContainer($path)
      * @return array|bool
      */
-    public function getMetaData($path)
+    public function getMetaData($path, $isNew = false)
```
Changing the signature of a public method of a non-final class is a technical breaking change (as derivative classes could have extended the method). There are no derivatives of this method in core and it was only added in 3.0, so perhaps we can choose to ignore introducing a breaking change at this point in the release, or we could just re-evaluate why we're adding this and go with the easiest approach of just lowering the error level on the message?

smg6511 replied:
I'd argue that's completely a technicality, as a default value is given for the new property and it's _only_ used to determine whether the thrown error is to be expected or not.

opengeek replied:
I'm with Mark on this one. I'd rather see the error level reduced here. The isNew parameter just doesn't even feel right on the getMetaData() method.

smg6511 replied:
I agree, It's not right, nor is how this method is leaned on to figure out the writability of a file. Given that I see _this_ PR's fix as a temporary bridge to the real and hopefully much more complete fix I'm working on, it's not a big deal changing this the way you're requesting. Will update later today...

smg6511 replied:
OK, the requested change was made. All - please be aware that this PR _does not_ fix the problems with saving non-media source based files.

Mark-H commented:
It took me 30 minutes to parse through all the discussion here and elsewhere and changes being proposed, but the key fix here is the one line change from `$sourceFile = $this->getSourceFile();` to `$sourceFile = $this->getStaticFileName();`. 

That prevents doubling up the path in isStaticSourceMutable, which fixes saving static elements. That seems like a logical patch, though incomplete for all currently known issues as indicated in the PR.

The error message is a red herring and IMO best just muted by lowering its level to WARN to avoid introducing a breaking signature change.

smg6511 commented:
@Mark-H - Yes, quite incomplete. Just note that I'm being pushed (for better or worse) to execute changes in smaller chunks to just be able to get them reviewed, which is why this PR addresses only a _part_ of the problem.

Commit:
Requested updates

Commit:
Update modMediaSource.php

Update error to warning in getMetaData ... forgot to do that in previous commit.

opengeek commented on core/lexicon/en/file.inc.php:
```diff
@@ -73,6 +73,7 @@
 $_lang['file_folder_remove'] = 'Delete Directory';
 $_lang['file_folder_remove_confirm'] = 'Are you sure you want to delete the directory: "[[+directory]]"?<br />This could potentially break your website.';
 $_lang['file_folder_rename'] = 'Rename Directory';
+$_lang['file_info_metadata_new'] = 'Unable to access metaData for the file “[[+file]]” because it is for a new element and has not been created yet. An attempt will be made to create the file automatically.';
```
This isn't still necessary since the isNew param was removed, correct?

smg6511 replied:
Yes, forgot that! Just removed it in new commit.

Commit:
Update file.inc.php

Remove unused entry.
